File: leetcode/cc/0021_merge_two_lists.cc
#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *merge_two_lists(ListNode *l1, ListNode *l2) {
    ListNode ll;
    ListNode *cur = &ll;

    while (l1 && l2) {
      if (l1->val < l2->val) {
        cur->next = new ListNode(l1->val);
        l1 = l1->next;
      } else {
        cur->next = new ListNode(l2->val);
        l2 = l2->next;
      }
      cur = cur->next;
    }

    if (l1)
      cur->next = l1;
    else
      cur->next= l2;
    return ll.next;
  }
};

void print_ll(ListNode *ll) {
  // iterate over merged linked list
  cout << "List:" << endl;
  while (ll) {
    cout << ll->val;
    ll = ll->next;
    if (ll != nullptr)
      cout << ", ";
  }
  cout << endl << endl;;
}

int get_rand(int beg, int end) {
  return rand() % ((end - beg) + 1) + beg;
}

void create_ll(ListNode *ll) {
  ListNode *cur = ll;
  int min = 0;
  int max = 10;
  int tmp = 0;
  int r = get_rand(0, 100);
  for (int i = 0; i < r; i++) {
    tmp = get_rand(tmp, max);
    cur->next = new ListNode(tmp);
    cur = cur->next;
    max += 10;
  }
}

int main() {
  // init rand generator
  srand(static_cast<int>(time(0)));

  // testcase
  ListNode l1;
  ListNode l2;

  // create linked lists
  create_ll(&l1);
  print_ll(&l1);
  create_ll(&l2);
  print_ll(&l2);

  // run solution
  Solution sol;
  ListNode *lmerged = sol.merge_two_lists(&l1, &l2);

  cout << "Merged Linked List:" << endl;
  print_ll(lmerged);

  return 0;
}

File: leetcode/go/0021_merge_two_lists.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	var cur *ListNode = &ListNode{}
	var head *ListNode = cur

	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			cur.Next = l1
			l1 = l1.Next
		} else {
			cur.Next = l2
			l2 = l2.Next
		}
		cur = cur.Next
	}

	if l1 != nil {
		cur.Next = l1
	} else {
		cur.Next = l2
	}
	return head.Next
}

func printll(ll *ListNode) {
	for ll != nil {
		fmt.Println(ll.Val)
		ll = ll.Next
	}
}

func getRand(beg int, end int) int {
	s1 := rand.NewSource(time.Now().UnixNano())
	r1 := rand.New(s1)
	return r1.Intn(end)%((end-beg)+1) + beg
}

func createll(ll *ListNode) {
	var cur *ListNode = ll
	var (
		max int = 10
		tmp int = 0
		r   int = getRand(0, 20)
	)

	for i := 0; i < r; i++ {
		tmp = getRand(tmp, max)
		cur.Next = new(ListNode)
		cur.Next.Val = tmp
		cur = cur.Next
		max += 10
	}
}

func main() {
	// testcase
	var (
		l1 ListNode
		l2 ListNode
	)

	// create linked lists
	createll(&l1)
	printll(&l1)
	createll(&l2)
	printll(&l2)
	var lmerged *ListNode = mergeTwoLists(&l1, &l2)

	// print result
	fmt.Printf("\nMerged:\n")
	printll(lmerged)
}
